---
title: "s9-3-1 mTOR and TSS isoforms"
author: "Yoichiro Sugimoto"
date: "`r format(Sys.time(), '%d %B, %Y')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
   github_document:
     toc: yes
     toc_depth: 2
     fig_width: 5
     fig_height: 5
---


# Environment setup and data preprocessing

```{r load_libraries, message = FALSE, warning = FALSE}

## Specify the number of CPUs to be used
processors <- 8

## library("BiocParallel")
## register(MulticoreParam(processors))

temp <- sapply(list.files("../functions", full.names = TRUE), source)
temp <- sapply(list.files("./functions", full.names = TRUE), source)  
source(file.path("../s6-differential-expression-and-tss-usage/functions/load_total_analysis_results.R"), chdir = TRUE)
source(file.path("../s8-analysis-of-translation/functions/test_differential_translation-v2.R"))

s4.tss.dir <- file.path(results.dir, "s4-tss-definition-and-tx-assignment")
s4.2.tx.assignment.dir <- file.path(s4.tss.dir, "s4-2-transcript-assignment")
s4.2.1.tss.tx.map.RCC4.dir <- file.path(s4.2.tx.assignment.dir, "s4-2-1-tss-transcript-mapping-RCC4")

s8.dir <- file.path(results.dir, "s8-analysis-of-translation")
s8.1.dir <- file.path(s8.dir, "s8-1-differentially-translated-mRNAs")
s8.1.1.dir <- file.path(s8.1.dir, "gene-level-dte")
s8.1.2.dir <- file.path(s8.1.dir, "tx-level-dte")
s8.2.dte.iso.dir <- file.path(s8.dir, "s8-2-differentially-translated-isoforms")
s8.3.dir <- file.path(s8.dir, "s8-3-validation-of-method")

s9.dir <- file.path(results.dir, "s9-integrative-analysis")

sq.dir <- file.path(results.dir, "sq-for-publication")
source.data.dir <- file.path(sq.dir, "sq1-source-data")
source.data.by.panel.dir <- file.path(source.data.dir, "by_panel")

set.seed(0)

```

# Data import


```{r data_import}

all.filtered.tss.dt <- file.path(
    s8.3.dir,
    "filtered_tss_for_polysome_analysis.csv"
) %>%
    fread


rcc4.vhl.mrl.dt <- file.path(
    s8.1.2.dir,
    "RCC4_VHL_EIF4E2_yy_xx__Torin1_vs_NA-mean_ribosome_loading.csv"
) %>%
    fread

tx.meta.info.dt <- file.path(s8.3.dir, "processed-tx-meta-info.csv") %>%
    fread

sig.rcc4.vhl.mrl.dt <- merge(
    rcc4.vhl.mrl.dt[
        biotype == "protein_coding",
        .(tss_name, gene_id, gene_name, MRL_treated, MRL_base, MRL_log2fc)],
    y = tx.meta.info.dt[
      , .(tss_name, TOP_motif_length, uORF_all, uORF_all_capped, cds_len)
    ],
    by = "tss_name"
)

sig.rcc4.vhl.mrl.dt <- sig.rcc4.vhl.mrl.dt[
        tss_name %in% all.filtered.tss.dt[
                      RCC4_VHL_NA == TRUE & RCC4_VHL_Torin1 == TRUE, tss_name
                  ]
]


```


# Analysis of TSS isoforms with differential sensitivity to mTOR in relation to the sequence features


```{r TSS_dependent_sensitivity_to_mTOR}

sig.rcc4.vhl.mrl.dt[, `:=`(
    max_MRL_log2fc = max(MRL_log2fc),
    min_MRL_log2fc = min(MRL_log2fc)
), by = gene_id]

sig.rcc4.vhl.mrl.dt[, `:=`(
    TSS_mTOR_group = case_when(
        MRL_log2fc == max_MRL_log2fc ~ "Resistant",
        MRL_log2fc == min_MRL_log2fc ~ "Sensitive"
    ),
    dMRL_log2fc = max_MRL_log2fc - min_MRL_log2fc
)]

sl.sig.rcc4.vhl.mrl.dt <- sig.rcc4.vhl.mrl.dt[
    !is.na(TSS_mTOR_group) & abs(dMRL_log2fc) > 0
][order(gene_name, TSS_mTOR_group)]

te.diff.classes <- c(
    "Small (1 ~ 1.2)", "Medium (1.2 ~ 1.5)", "Large (> 1.5)"
)

sl.sig.rcc4.vhl.mrl.dt[, `:=`(
    te_diff_by_tss = case_when(
        abs(dMRL_log2fc) > log2(1.5) ~ "Large (> 1.5)",
        abs(dMRL_log2fc) > log2(1.2) ~ "Medium (1.2 ~ 1.5)",
        abs(dMRL_log2fc) > 0 ~ "Small (1 ~ 1.2)"
    ) %>% factor(levels = te.diff.classes)
)]

d.sl.sig.rcc4.vhl.mrl.dt <- dcast(
    sl.sig.rcc4.vhl.mrl.dt,
    gene_id + gene_name + te_diff_by_tss ~ TSS_mTOR_group,
    value.var = c(
        "tss_name", "MRL_log2fc", "TOP_motif_length", "uORF_all", "cds_len"
    )
)

print("All genes with annotation of multiple TSS available")
d.sl.sig.rcc4.vhl.mrl.dt[, table(te_diff_by_tss) %>% addmargins]

## Test function
runWilcox <- function(te.diff.class, d.sl.sig.rcc4.vhl.mrl.dt, sl.genes, test.col){
    all.dt <- d.sl.sig.rcc4.vhl.mrl.dt[gene_id %in% sl.genes] 
    
    test.dt <- all.dt[te_diff_by_tss == te.diff.class] 
    
    wil.p <-  test.dt %>%
    {wilcox.test(
         .[, get(paste0(test.col, "_Sensitive"))],
         .[, get(paste0(test.col, "_Resistant"))],
         alternative = "two.sided",
         paired = TRUE
     )$p.value}

    wil.res.dt <- data.table(
        tested_data = test.col,
        te_diff_by_tss = te.diff.class,
        N = nrow(test.dt),
        all_N = nrow(all.dt),
        wilcox_p = wil.p
    )
    return(wil.res.dt)
}

sig.th <- 0.05

## TOP mptof length
print("TOP motif length")
diff.top.genes <- d.sl.sig.rcc4.vhl.mrl.dt[
  TOP_motif_length_Sensitive != TOP_motif_length_Resistant, gene_id
]

top.test.res.dt <- lapply(
    te.diff.classes,
    runWilcox,
    d.sl.sig.rcc4.vhl.mrl.dt = d.sl.sig.rcc4.vhl.mrl.dt,
    sl.genes = diff.top.genes,
    test.col = "TOP_motif_length"
) %>%
    rbindlist

top.test.res.dt[, padj := p.adjust(wilcox_p, method = "holm")]
top.test.res.dt[, sig_mark := case_when(
                      padj < sig.th * 0.1 ~ "**",
                      padj < sig.th ~ "*",
                      TRUE ~ NA_character_
                  )
                ]
print(top.test.res.dt)

ggplot(
    data = sl.sig.rcc4.vhl.mrl.dt[gene_id %in% diff.top.genes],
    aes(
        x = te_diff_by_tss,
        y = TOP_motif_length
    )
) +
    geom_boxplot(aes(fill = TSS_mTOR_group), outlier.shape = NA) +
    theme(aspect.ratio = 1.5) +
    scale_fill_bright(name = "Sensitivity of isoform to mTOR inhibition") +
    scale_x_discrete(guide = guide_axis(angle = 90)) +
    coord_cartesian(ylim = c(0, 6.5)) +
    xlab("MRL log2 fold change difference between TSS isoforms") +
    ylab("TOP motif length")

## uORF length
print("uORF length")

diff.uorf.genes <- d.sl.sig.rcc4.vhl.mrl.dt[
  uORF_all_Sensitive != uORF_all_Resistant, gene_id
]

uorf.test.res.dt <- lapply(
    te.diff.classes,
    runWilcox,
    d.sl.sig.rcc4.vhl.mrl.dt = d.sl.sig.rcc4.vhl.mrl.dt,
    sl.genes = diff.uorf.genes,
    test.col = "uORF_all"
) %>%
    rbindlist

uorf.test.res.dt[, padj := p.adjust(wilcox_p, method = "holm")]
uorf.test.res.dt[, sig_mark := case_when(
                      padj < sig.th * 0.1 ~ "**",
                      padj < sig.th ~ "*",
                      TRUE ~ NA_character_
                  )
                ]
print(uorf.test.res.dt)

ggplot(
    data = sl.sig.rcc4.vhl.mrl.dt[gene_id %in% diff.uorf.genes],
    aes(
        x = te_diff_by_tss,
        y = uORF_all,
        fill = TSS_mTOR_group
    )
) +
    geom_boxplot(outlier.shape = NA) +
    theme(aspect.ratio = 1.5) +
    scale_fill_bright(name = "Sensitivity of isoform to mTOR inhibition") +
    scale_x_discrete(guide = guide_axis(angle = 90)) +
    coord_cartesian(ylim = c(0, 7)) +
    xlab("MRL log2 fold change difference between TSS isoforms") +
    ylab("uORF number")

source.data.base.cols <- c("gene_id", "gene_name", "tss_name_Sensitive", "tss_name_Resistant", "MRL_log2fc_Sensitive", "MRL_log2fc_Resistant")

temp <- exportSourceData(
    dt = d.sl.sig.rcc4.vhl.mrl.dt[
        gene_id %in% c(diff.top.genes, diff.uorf.genes)
    ],
    original.colnames = c(
        source.data.base.cols,
        "TOP_motif_length_Sensitive", "TOP_motif_length_Resistant",
        "uORF_all_Sensitive", "uORF_all_Resistant"
    ),
    export.colnames = c(
        source.data.base.cols,
        "TOP motif length (Sensitive)", "TOP motif length (Resistant)",
        "uORF number (Sensitive)", "uORF number (Resistant)"
    ),
    export.file.name = "Extended Data Fig. 5d.csv"
)

```


# Session information

```{r sessionInfo}

sessionInfo()

```
