---
title: "s9-3-1 Appendix I for s-9-1-2"
author: "Yoichiro Sugimoto"
date: "`r format(Sys.time(), '%d %B, %Y')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
   html_document:
     highlight: haddock
     toc: yes
     toc_depth: 2
     keep_md: yes
     fig_width: 5
     fig_height: 5
---


# Overview

Extract n values.

```{r load libraries, message = FALSE, warning = FALSE}

## Specify the number of CPUs to be used
processors <- 8

temp <- sapply(list.files("../functions", full.names = TRUE), source)
temp <- sapply(list.files("../s8-analysis-of-translation/functions", full.names = TRUE), source, chdir = TRUE)
temp <- sapply(list.files("./functions", full.names = TRUE), source, chdir = TRUE)
source("../s6-differential-expression-and-tss-usage/functions/load_total_analysis_results.R", chdir = TRUE)

set.seed(0)

sig.th <- 0.05

```


```{r define directory}

annot.dir <- normalizePath(file.path("../../annotation/"))
annot.ps.dir <- file.path(annot.dir, "hg38_annotation/processed_data/")
annot.R.file <- list.files(
    annot.ps.dir,
    pattern = glob2rx("*primary_transcript_annotation*.rdata"),
    full.names = TRUE
)
load(annot.R.file)

results.dir <- file.path("../../results")
s4.tss.dir <- file.path(results.dir, "s4-tss-definition-and-tx-assignment")
s4.1.tss.def.dir <- file.path(s4.tss.dir, "s4-1-tss-definition")
s4.1.7.count.per.tss.dir <- file.path(s4.1.tss.def.dir, "s4-1-7-count-per-tss")
s4.3.tx.info.dir <- file.path(s4.tss.dir, "s4-3-transcript-info")
s4.3.1.tx.info.rcc4.dir <- file.path(s4.3.tx.info.dir, "s4-3-1-transcript-info-for-RCC4")


s6.dir <- file.path(results.dir, "s6-differential-regulation-analysis")
s6.1.dir <- file.path(s6.dir, "s6-1-differentially-expressed-genes")

s8.dir <- file.path(results.dir, "s8-analysis-of-translation")
s8.1.dir <- file.path(s8.dir, "s8-1-differentially-translated-mRNAs")
s8.1.1.dir <- file.path(s8.1.dir, "gene-level-dte")
s8.1.2.dir <- file.path(s8.1.dir, "tx-level-dte")
s8.3.dir <- file.path(s8.dir, "s8-3-validation-of-method")

## s9.dir <- file.path(results.dir, "s9-integrative-analysis")

sample.file <- file.path("../../data/sample_data/processed_sample_file.csv")
sample.dt <- fread(sample.file)
sample.names <- sample.dt[, sample_name]

```

# Analysis of the association of TOP motif length and changes in translation upon VHL loss


## QC for the TOP motif length assignment


```{r qc for top motif length}

tx.meta.data.rcc4.dt <- file.path(
    s4.3.1.tx.info.rcc4.dir,
    "transcript-meta-information-RCC4-VHL.csv"
) %>% fread

tx.meta.data.786o.dt <- file.path(
    s4.3.tx.info.dir, "s4-3-2-transcript-info-for-786O",
    "transcript-meta-information-786O-VHL.csv"
) %>% fread


## QC
r4.temp.dt <- copy(tx.meta.data.rcc4.dt)
setnames(r4.temp.dt, old = "tss_p1_pTOP", new = "TOP_in_RCC4")

c7o.temp.dt <- copy(tx.meta.data.786o.dt)
setnames(c7o.temp.dt, old = "tss_p1_pTOP", new = "TOP_in_786O")

temp.qc.dt <- merge(
    r4.temp.dt[, .(tss_name, TOP_in_RCC4)],
    c7o.temp.dt[, .(tss_name, TOP_in_786O)],
    by = "tss_name"
)

ggplot(
    data = temp.qc.dt,
    aes(
        x = TOP_in_RCC4,
        y = TOP_in_786O
    )
) +
    geom_point()


```

## Analysis


```{r top motif length vs VHL loss trsl}

all.filtered.tss.dt <- file.path(
    s8.3.dir,
    "filtered_tss_for_polysome_analysis.csv"
) %>% fread

s8.3.dir <- file.path(s8.dir, "s8-3-validation-of-method")

mergeTranslationAndMetaData <- function(input.diff.trsl.file, filtered.tss.names, tx.meta.data.dt, cell.name){
    diff.trsl.dt <- file.path(
        s8.1.2.dir,
        input.diff.trsl.file
    ) %>% fread

    diff.trsl.dt <- diff.trsl.dt[
        tss_name %in% filtered.tss.names 
    ]

    diff.trsl.dt <- merge(
        diff.trsl.dt,
        tx.meta.data.dt[, .(tss_name, tss_p1_pTOP)],
        by = "tss_name"
    )
    
    diff.trsl.dt <- diff.trsl.dt[
        !is.na(tss_p1_pTOP) & !is.na(MRL_log2fc)
    ]

    diff.trsl.dt[, `:=`(
        rd_TOP_motif_length = round(tss_p1_pTOP) %>%
            {if_else(. >= 8, "8+", as.character(.))},
        cell_name = cell.name
    )]

    diff.trsl.dt[, mRNA_N := .N, by = list(rd_TOP_motif_length)]

    return(diff.trsl.dt)
}


r4.filtered.tx <- all.filtered.tss.dt[
    (RCC4_noVHL_NA == TRUE) & (RCC4_VHL_NA == TRUE), tss_name
]

c7.filtered.tx <- all.filtered.tss.dt[
    (c786O_noVHL_EIF4E2_yy_NA == TRUE) & (c786O_VHL_EIF4E2_yy_NA == TRUE), tss_name
]


top.vhl.loss.dt <- mapply(
    FUN = mergeTranslationAndMetaData,
    input.diff.trsl.file = c("RCC4_xx_EIF4E2_yy_NA__noVHL_vs_VHL.csv",
      "786O_xx_EIF4E2_yy_NA__noVHL_vs_VHL.csv"),
    filtered.tss.names = list(r4.filtered.tx, c7.filtered.tx),
    tx.meta.data.dt = list(tx.meta.data.rcc4.dt, tx.meta.data.786o.dt),
    cell.name = c("RCC4", "786-O"),
    SIMPLIFY = FALSE
) %>%
    rbindlist

top.vhl.loss.sig.dt <- lapply(
    c("RCC4", "786-O"),
    function(x){
        data.table(
            cell_name = x,
            top.vhl.loss.dt[cell_name == x] %$%
            pairwise.wilcox.test(
                x = MRL_log2fc,
                g = rd_TOP_motif_length,
                p.adjust.method = "none",
                alternative = "two.sided"
            )$p.value[, 1] %>% stack %>% data.table            
        )
    }
) %>% rbindlist %>%
{.[, padj := p.adjust(values, method = "holm")]} %>%
{.[, `:=`(
     sig_mark = case_when(
         padj < sig.th * 0.1 ~ "**",
         padj < sig.th ~ "*",
         TRUE ~ NA_character_
     ),
     rd_TOP_motif_length = ind
 )]} %>%
merge(
    y = top.vhl.loss.dt[, .N, by = list(cell_name, rd_TOP_motif_length)],
    by = c("cell_name", "rd_TOP_motif_length"),
    all.y = TRUE
) %>%
{.[
   , cell_name := factor(cell_name, levels = c("RCC4", "786-O"))
 ][order(cell_name)]} %T>%
print

merge(
    top.vhl.loss.dt,
    top.vhl.loss.sig.dt,
    by = c("cell_name", "rd_TOP_motif_length"),
    all.x = TRUE
) %>% {.[, cell_name := factor(cell_name, levels = c("RCC4", "786-O"))]} %>%
    ggplot(
        aes(
            x = rd_TOP_motif_length,
            y = MRL_log2fc,
            color = mRNA_N < 10
        )
    ) +
    geom_hline(yintercept = 0, color = "gray60") +
    geom_boxplot(outlier.shape = NA) +
    stat_summary(
        geom = 'text', aes(label = sig_mark),
        fun = function(x){boxplot.stats(x)$stats[5]}, 
        vjust = -0.8, color = "black", size = 6
    ) +
    facet_grid(cell_name ~ .) +
    theme(
        aspect.ratio = 1,
        legend.position = "none"
    ) +
    scale_color_manual(values = c("TRUE" = "gray60", "FALSE" = "black")) +
    xlab("TOP motif length") +
    ylab("MRL log2 fold change upon VHL loss")


```


# Session information

```{r sessionInfo}

sessionInfo()

```
