---
title: "s7-2 Define HIF1A and HIF2A binding ratio"
author: "Yoichiro Sugimoto"
date: "`r format(Sys.time(), '%d %B, %Y')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
   github_document:
     toc: yes
     toc_depth: 2
     fig_width: 5
     fig_height: 5
---


# Overview

Based on ChIP-Seq data, HIF binding sites and the ratio of HIF2A binding relative to that of HIF1A will be defined.


# Environment setup and data preprocessing

```{r load_libraries, message = FALSE, warning = FALSE}

## Bioconductor
library("GenomicFeatures")
library("DiffBind")
## rmarkdown
library("knitr")
library("kableExtra")

## Specify the number of CPUs to be used
processors <- 8

## library("BiocParallel")
## register(MulticoreParam(processors))

temp <- sapply(list.files("../functions", full.names = TRUE), source)

results.dir <- file.path("../../results")

s4.tss.dir <- file.path(results.dir, "s4-tss-definition-and-tx-assignment")
s4.1.tss.def.dir <- file.path(s4.tss.dir, "s4-1-tss-definition")
s4.1.6.filtered.tss.dir <- file.path(s4.1.tss.def.dir, "s4-1-6-filtered-tss")
s4.3.tx.info.dir <- file.path(s4.tss.dir, "s4-3-transcript-info")
s4.3.1.tx.info.rcc4.dir <- file.path(s4.3.tx.info.dir, "s4-3-1-transcript-info-for-RCC4")

s7.dir <- file.path(results.dir, "s7-HIF-binding-site")
s7.1.dir <- file.path(s7.dir, "s7-1-ChIP-Seq")
s7.1.macs.dir <- file.path(s7.1.dir, "macs_out")
s7.1.hre.bed.dir <- file.path(s7.1.macs.dir, "IDR_filtered_macs_HRE_bed")
s7.1.hre.narrowPeaks.dir <- file.path(s7.1.macs.dir, "IDR_filtered_macs_HRE_narrowPeaks")
s7.2.hif1.2a.dir <- file.path(s7.dir, "s7-2-HIF1A-and-HIF2A-ratio")

create.dirs(c(
    s7.2.hif1.2a.dir
))

set.seed(0)

```


# Define HIF binding sites and the ratio of HIF2A/HIF1A binding


```{r positional_relationship_of_novel_tss_and_hif_binding_sites}

reprocessed.chip.seq.data.dir <- file.path(
    "../../../20200519_reprocessing_HIF_ChIP-Seq"
)

chip.seq.sample.dt <- file.path(
    "../../data/sample_data/20210203_ChIPSeq_sample_data.csv"
) %>% fread

chip.seq.sample.dt[, `:=`(
    bamReads = file.path(
        reprocessed.chip.seq.data.dir, "processed_data/sorted_bam", bamReads
    ),
    Peaks = file.path(
        s7.1.hre.narrowPeaks.dir, Peaks
    )
)]

all.chip.peak.gr <- dba(sampleSheet = chip.seq.sample.dt) %>%
    dba.count %>%
    dba.contrast(categories = DBA_FACTOR, minMembers = 2) %>%
    dba.analyze(bFullLibrarySize = FALSE) %>%
    dba.report(th = 1)

readChIPSeqData <- function(chip.seq.core.name){
    idr.narrowpeak.file <- file.path(
        s7.1.hre.narrowPeaks.dir,
        paste0(chip.seq.core.name, ".narrowPeak")
    )
    temp.idr.narrowpeak.dt <- fread(idr.narrowpeak.file)
    
    idr.narrowpeak.dt <- data.table(
        chr = temp.idr.narrowpeak.dt[, V1],
        start = temp.idr.narrowpeak.dt[, V2 + V10] + 1, # bed is 0-base
        end = temp.idr.narrowpeak.dt[, V2 + V10], 
        enrichment = temp.idr.narrowpeak.dt[, V7],
        mlog10qval = temp.idr.narrowpeak.dt[, V9],
        peak = temp.idr.narrowpeak.dt[, V2 + V10] + 1 # bed is 0-base
    )

    chip.seq.summit.gr <- makeGRangesFromDataFrame(
        idr.narrowpeak.dt, keep.extra.columns = TRUE
    )
}

hif1a.chip.seq.summit.gr <- readChIPSeqData("RCC4_N_HIF1A")
hif2a.chip.seq.summit.gr <- readChIPSeqData("RCC4_N_HIF2A")

hif1a.ol.dt <- findOverlaps(
    hif1a.chip.seq.summit.gr,
    all.chip.peak.gr,
) %>% as.data.frame %>% data.table

hif2a.ol.dt <- findOverlaps(
    hif2a.chip.seq.summit.gr,
    all.chip.peak.gr
) %>% as.data.frame %>% data.table

mcols(all.chip.peak.gr)[["index"]] <- 1:length(all.chip.peak.gr)

all.chip.peak.mcols.dt <- all.chip.peak.gr %>%
    as.data.frame %>% data.table

np.cols <- c("enrichment", "mlog10qval", "peak")

h1.mcol.dt <- cbind(
    all.chip.peak.mcols.dt[hif1a.ol.dt[, subjectHits], "index", with = FALSE],
    mcols(hif1a.chip.seq.summit.gr[hif1a.ol.dt[, queryHits]]) %>% as.data.frame %>% data.table
)
setnames(h1.mcol.dt, old = np.cols, new = paste0("HIF1A_", np.cols))

h2.mcol.dt <- cbind(
    all.chip.peak.mcols.dt[hif2a.ol.dt[, subjectHits], "index", with = FALSE],
    mcols(hif2a.chip.seq.summit.gr[hif2a.ol.dt[, queryHits]]) %>% as.data.frame %>% data.table
)
setnames(h2.mcol.dt, old = np.cols, new = paste0("HIF2A_", np.cols))

all.chip.peak.comp.mcols.dt <- Reduce(
    function(...) merge(..., all = TRUE, by = "index"),
    list(
        all.chip.peak.mcols.dt,
        h1.mcol.dt,
        h2.mcol.dt
    )) %>%
    {.[order(index, HIF1A_enrichment, HIF2A_enrichment)][!duplicated(index)]}

all.chip.peak.comp.mcols.dt[, `:=`(
    peak_position = case_when(
        is.na(HIF2A_peak) & !is.na(HIF1A_peak) ~ HIF1A_peak,
        is.na(HIF1A_peak) & !is.na(HIF2A_peak) ~ HIF2A_peak,
        ## HIF1A_mlog10qval >= HIF2A_mlog10qval ~ HIF1A_peak,
        Fold >= 0 ~ HIF1A_peak,
        Fold < 0 ~ HIF2A_peak
    )
)]

fwrite(
    all.chip.peak.comp.mcols.dt,
    file.path(
        s7.2.hif1.2a.dir,
        "HIF1A-HIF2A-binding-summary.csv"
    )
)

```

## Export HIF binding sites in bed format


```{r export_HIF_binding_sites_information}

hif.bed.dt <- all.chip.peak.comp.mcols.dt[, list(
    chrom = seqnames,
    start = start - 1,
    end = end,
    name = paste0("peak_", index, "_HIF2A/HIF1_log2fc_", - Fold),
    score = 1,
    strand = strand,
    thick_start = peak_position - 1,
    thick_end = peak_position,
    itemRgb = "0,0,255",
    block_count = 1,
    blockSizes = end - start - 1,
    blockStart = 0    
)]

fwrite(
    hif.bed.dt,
    file.path(
        s7.2.hif1.2a.dir,
        "HIF-binding-site.bed"
    ),
    col.names = FALSE, sep = "\t"
)


```



# Assign closest HIF binding sites to all TSS


```{r assign_closest_hif_binding_sites_to_TSS}

filtered.tss.with.quantile.dt <- file.path(
  s4.1.6.filtered.tss.dir,
  "filtered-tss-with-quantile.csv"
) %>% fread

filtered.tss.with.quantile.dt <- filtered.tss.with.quantile.dt[gene_id != ""]

filtered.tss.with.quantile.gr <- makeGRangesFromDataFrame(
    filtered.tss.with.quantile.dt,
    keep.extra.columns = TRUE
)

chip.seq.summit.gr <- makeGRangesFromDataFrame(
    all.chip.peak.comp.mcols.dt[!is.na(peak_position)],
    keep.extra.columns = TRUE
)

start(chip.seq.summit.gr) <- mcols(chip.seq.summit.gr)[["peak_position"]]
end(chip.seq.summit.gr) <- mcols(chip.seq.summit.gr)[["peak_position"]] 

nearest.hif.dt <- nearest(
    filtered.tss.with.quantile.gr,
    chip.seq.summit.gr,
    ignore.strand = TRUE,
    select = "all"
) %>% as.data.frame %>% data.table

tss.hif.pos.dt <- cbind(
    filtered.tss.with.quantile.dt[nearest.hif.dt[, queryHits]],
    as.data.frame(chip.seq.summit.gr[nearest.hif.dt[, subjectHits]]) %>%
    data.table %>% {
        .[, list(
            hre_position = start,
            Conc_HIF1A, Conc_HIF2A, Fold
        )]
    }
)

tss.hif.pos.dt[, `:=`(
    tss_pos_to_hif = ifelse(strand == "+", 1 , -1) * (q50 - hre_position)
)]
tss.hif.pos.dt[, `:=`(
    log10_tss_pos_to_hif = sign(tss_pos_to_hif) * log10(abs(tss_pos_to_hif))
)]
tss.hif.pos.dt <- tss.hif.pos.dt[gene_id != ""]

tss.hif.pos.dt[, `:=`(
    HIF2A_enrichment_with_dist_th = if_else(
        abs(tss_pos_to_hif) < 500 * 10^3, - Fold, NA_real_
    )
)]

fwrite(
    tss.hif.pos.dt,
    file.path(
        s7.2.hif1.2a.dir,
        "filtered-tss-and-nearest-hif-binding-position.csv"
    )
)

```


# session information

```{r sessionInfo}

sessionInfo()

```
