---
title: "s8-1-2 Trasncript level translationa analysis"
author: "Yoichiro Sugimoto"
date: "`r format(Sys.time(), '%d %B, %Y')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
   html_document:
     highlight: haddock
     toc: yes
     toc_depth: 2
     keep_md: yes
     fig_width: 5
     fig_height: 5
---


# Overview

mRNA isoform level changes in translation upon VHL loss or Torin 1 treatment will be examined.


```{r load libraries, message = FALSE, warning = FALSE}

## Data analysis
library("tidymodels")
library("matrixStats")
## Data visualization
library("ggrepel")
## Specify the number of CPUs to be used
processors <- 8
library("BiocParallel")
register(MulticoreParam(processors))

temp <- sapply(list.files("../functions", full.names = TRUE), source)
temp <- sapply(list.files("./functions", full.names = TRUE), source, chdir = TRUE)

set.seed(0)

```


```{r define directory}

## sample.file <- file.path("../../data/sample_data/processed_sample_file.csv")
annot.dir <- normalizePath(file.path("../../annotation/"))
annot.ps.dir <- file.path(annot.dir, "hg38_annotation/processed_data/")
annot.R.file <- list.files(
    annot.ps.dir,
    pattern = glob2rx("*primary_transcript_annotation*.rdata"),
    full.names = TRUE
)
load(annot.R.file)

results.dir <- file.path("../../results")

s3.alignment.stats.dir <- file.path(results.dir, "s3-alignment-statistics")
s3.4.poly.size.factor.dir <- file.path(s3.alignment.stats.dir, "polysome_size_factor")

s4.tss.dir <- file.path(results.dir, "s4-tss-definition-and-tx-assignment")
s4.1.tss.def.dir <- file.path(s4.tss.dir, "s4-1-tss-definition")
s4.1.7.count.per.tss.dir <- file.path(s4.1.tss.def.dir, "s4-1-7-count-per-tss")

s8.dir <- file.path(results.dir, "s8-analysis-of-translation")
s8.1.dir <- file.path(s8.dir, "s8-1-differentially-translated-mRNAs")
s8.1.2.dir <- file.path(s8.1.dir, "tx-level-dte")

create.dirs(c(
  s8.dir,
  s8.1.dir,
  s8.1.2.dir
))

```

# Read raw and normalize count table

```{r read start count data}

sample.names <- poly.sample.dt[, sample_name]

count.per.tss.file <- file.path(s4.1.7.count.per.tss.dir, "count-per-confident-tss.csv")
count.per.tss.dt <- fread(count.per.tss.file)

## Only polysome proflied sequence data are eported
poly.count.per.tss.dt <- count.per.tss.dt[
  , c("tss_name", grep("polysome_", colnames(count.per.tss.dt), value = TRUE)),
    with = FALSE
]

colnames(poly.count.per.tss.dt) <- colnames(poly.count.per.tss.dt) %>%
    gsub("polysome_", "", .)

poly.count.per.tss.dt[, gene_id := str_split_fixed(tss_name, "_", n = 2)[, 1]]

## Only analyze transcripts from genes in primary.tx.dt
poly.count.per.tss.dt <- poly.count.per.tss.dt[
    gene_id %in% unique(primary.tx.dt[, gene_id])
]


## Read normalization factor
poly.sizefactor.dt <- fread(
    file = file.path(
        s3.4.poly.size.factor.dir,
        "library_size_factor_by_ERCC.csv"
    )
)

poly.sizefactor.with.ribo0.dt <- fread(
    file = file.path(
        s3.4.poly.size.factor.dir,
        "library_size_factor_with_ribo0_by_ERCC.csv"
    )
)

```

# Data pre-processing


```{r data preprocessing}

ref.column.name <- "tss_name"
annot.cols <- c("gene_id", "gene_name", "biotype")

poly.count.per.tss.with.meta.dt <- merge(
    primary.tx.dt[!duplicated(gene_id), .(gene_id, gene_name, biotype)],
    poly.count.per.tss.dt,
    by = "gene_id"
)

poly.tss.count.dt <- cbind(
  poly.count.per.tss.with.meta.dt[, c(ref.column.name, annot.cols), with = FALSE], 
  poly.count.per.tss.with.meta.dt[, colnames(poly.count.per.tss.with.meta.dt) %in% poly.sample.dt[, sample_name], with = FALSE]
)

```


# Preparation of data necessary for running DESEeq2


```{r deseq2 prep for dte analysis}

countDt2Df <- function(count.dt, coldata.df, ref.column.name = "tss_name", annot.cols){
    count.df <- as.data.frame(count.dt)
    rownames(count.df) <- count.df[, ref.column.name]
    annot.df <- count.df[, c(ref.column.name, annot.cols)]
    count.df <- count.df[, rownames(coldata.df)]
    return(list(count.df = count.df, annot.df = annot.df))
}

deseq2.poly.in.list <- countDt2Df(
    poly.tss.count.dt,
    poly.coldata.df,
    ref.column.name = ref.column.name, annot.cols = annot.cols
)

if(!all(colnames(deseq2.poly.in.list$count.df) == rownames(poly.coldata.df))){
    stop("colnames of count.df does not match with rownames of poly.coldata.df")
} 

annot.dt <- data.table(deseq2.poly.in.list$annot.df)
setkeyv(annot.dt, ref.column.name)

```

# Analysis of the effect of VHL

## RCC-4 noVHL vs RCC-4 VHL

```{r rcc4 diff translation}

## Select necessary data frame
sl.translation.comparison.dt <- translation.comparison.dt[1, ]
print("The following data will be analyzed:")
print(sl.translation.comparison.dt)

test.poly.coldata.df <- subsetColdata(sl.translation.comparison.dt[, comparison], poly.coldata.df)

temp <- analyzeDtg(
    count.df = deseq2.poly.in.list$count.df,
    annot.df = deseq2.poly.in.list$annot.df,
    ref.column.name = ref.column.name,
    comparison.name = sl.translation.comparison.dt[1, comparison],
    deseq2.formula = sl.translation.comparison.dt[, exp_formula][[1]],
    input.sample.data.df = test.poly.coldata.df,
    poly.sizefactor.dt = poly.sizefactor.dt,
    sizefactor.col = "size_factor",
    s8.1.dir = s8.1.2.dir,
    processors = processors
)


```

## RCC-4 noVHL Torin-1 vs no treatment

```{r rcc4 noVHL Torin1 diff translation}

## Select necessary data frame
sl.translation.comparison.dt <- translation.comparison.dt[2, ]
print("The following data will be analyzed:")
print(sl.translation.comparison.dt)

test.poly.coldata.df <- subsetColdata(sl.translation.comparison.dt[, comparison], poly.coldata.df)
test.poly.coldata.df <- test.poly.coldata.df[
    !grepl(
         "RCC4_noVHL_EIF4E2_NA_1_Torin1_(cyto|ribo[[:digit:]]|ribo[[:digit:]](A|B))_1",
         rownames(test.poly.coldata.df)
     ), ]

temp <- analyzeDtg(
    count.df = deseq2.poly.in.list$count.df,
    annot.df = deseq2.poly.in.list$annot.df,
    ref.column.name = ref.column.name,
    comparison.name = sl.translation.comparison.dt[1, comparison],
    deseq2.formula = sl.translation.comparison.dt[, exp_formula][[1]],
    input.sample.data.df = test.poly.coldata.df,
    poly.sizefactor.dt = poly.sizefactor.dt,
    sizefactor.col = "size_factor",
    s8.1.dir = s8.1.2.dir,
    processors = processors
)


```



## RCC-4 VHL treated with Torin-1 vs no treatment

```{r rcc4 VHL Torin1 diff translation}

## Select necessary data frame
sl.translation.comparison.dt <- translation.comparison.dt[3, ]
print("The following data will be analyzed:")
print(sl.translation.comparison.dt)

test.poly.coldata.df <- subsetColdata(sl.translation.comparison.dt[, comparison], poly.coldata.df)
test.poly.coldata.df <- test.poly.coldata.df[
    !grepl(
         "RCC4_VHL_EIF4E2_NA_1_Torin1_(cyto|ribo[[:digit:]]|ribo[[:digit:]](A|B))_1",
         rownames(test.poly.coldata.df)
     ), ]

temp <- analyzeDtg(
    count.df = deseq2.poly.in.list$count.df,
    annot.df = deseq2.poly.in.list$annot.df,
    ref.column.name = ref.column.name,
    comparison.name = sl.translation.comparison.dt[1, comparison],
    deseq2.formula = sl.translation.comparison.dt[, exp_formula][[1]],
    input.sample.data.df = test.poly.coldata.df,
    poly.sizefactor.dt = poly.sizefactor.dt,
    poly.sizefactor.with.ribo0.dt = poly.sizefactor.with.ribo0.dt,
    sizefactor.col = "size_factor",
    s8.1.dir = s8.1.2.dir,
    processors = processors
)


```



## 786-O noVHL vs 786-O VHL

```{r 786o diff translation}

## Select necessary data frame
sl.translation.comparison.dt <- translation.comparison.dt[4, ]
print("The following data will be analyzed:")
print(sl.translation.comparison.dt)

test.poly.coldata.df <- subsetColdata(sl.translation.comparison.dt[, comparison], poly.coldata.df)
## test.poly.coldata.df <- test.poly.coldata.df[test.poly.coldata.df$replicate %in% 1:3, ]

temp <- analyzeDtg(
    count.df = deseq2.poly.in.list$count.df,
    annot.df = deseq2.poly.in.list$annot.df,
    ref.column.name = ref.column.name,
    comparison.name = sl.translation.comparison.dt[1, comparison],
    deseq2.formula = sl.translation.comparison.dt[, exp_formula][[1]],
    input.sample.data.df = test.poly.coldata.df,
    poly.sizefactor.dt = poly.sizefactor.dt,
    sizefactor.col = "size_factor",
    s8.1.dir = s8.1.2.dir,
    processors = processors
)


```


# EIF4E2 dependent translation changes


## 786-O vs 786-O EIF4E2 KO

```{r 786O eif4e2 ko diff translation}

sl.translation.comparison.dt <- translation.comparison.dt[5, ]
print("The following data will be analyzed:")
print(sl.translation.comparison.dt)

test.poly.coldata.df <- subsetColdata(sl.translation.comparison.dt[, comparison], poly.coldata.df)

temp <- analyzeDtg(
    count.df = deseq2.poly.in.list$count.df,
    annot.df = deseq2.poly.in.list$annot.df,
    ref.column.name = ref.column.name,
    comparison.name = sl.translation.comparison.dt[1, comparison],
    deseq2.formula = sl.translation.comparison.dt[, exp_formula][[1]],
    input.sample.data.df = test.poly.coldata.df,
    poly.sizefactor.dt = poly.sizefactor.dt,
    sizefactor.col = "size_factor",
    s8.1.dir = s8.1.2.dir,
    processors = processors
)


```

## 786-O VHL vs 786-O VHL EIF4E2 KO


```{r 786O VHL eif4e2 ko diff translation}

sl.translation.comparison.dt <- translation.comparison.dt[6, ]
print("The following data will be analyzed:")
print(sl.translation.comparison.dt)

test.poly.coldata.df <- subsetColdata(sl.translation.comparison.dt[, comparison], poly.coldata.df)

temp <- analyzeDtg(
    count.df = deseq2.poly.in.list$count.df,
    annot.df = deseq2.poly.in.list$annot.df,
    ref.column.name = ref.column.name,
    comparison.name = sl.translation.comparison.dt[1, comparison],
    deseq2.formula = sl.translation.comparison.dt[, exp_formula][[1]],
    input.sample.data.df = test.poly.coldata.df,
    poly.sizefactor.dt = poly.sizefactor.dt,
    sizefactor.col = "size_factor",
    s8.1.dir = s8.1.2.dir,
    processors = processors
)

```


# Session information

```{r sessionInfo}

sessionInfo()

```
