---
title: "s8-3-2-1 Analysis of mTOR-dependent translational regulation (1/2)"
author: "Yoichiro Sugimoto"
date: "`r format(Sys.time(), '%d %B, %Y')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
   github_document:
     toc: yes
     toc_depth: 2
     fig_width: 5
     fig_height: 5
---


# Overview


The translational effect of mTOR inhibition in RCC-4 VHL will be examined here.


```{r load_libraries, message = FALSE, warning = FALSE}

## Specify the number of CPUs to be used
processors <- 8
## library("BiocParallel")
## register(MulticoreParam(processors))

temp <- sapply(list.files("../functions", full.names = TRUE), source)
source("../s6-differential-expression-and-tss-usage/functions/load_total_analysis_results.R", chdir = TRUE)

set.seed(0)

```


```{r define_directory}

sample.file <- file.path("../../data/sample_data/processed_sample_file.csv")

annot.dir <- normalizePath(file.path("../../annotation/"))
annot.ps.dir <- file.path(annot.dir, "hg38_annotation/processed_data/")
annot.R.file <- list.files(
    annot.ps.dir,
    pattern = glob2rx("*primary_transcript_annotation*.rdata"),
    full.names = TRUE
)
load(annot.R.file)

results.dir <- file.path("../../results")

s4.tss.dir <- file.path(results.dir, "s4-tss-definition-and-tx-assignment")
s4.1.tss.def.dir <- file.path(s4.tss.dir, "s4-1-tss-definition")
s4.1.6.filtered.tss.dir <- file.path(s4.1.tss.def.dir, "s4-1-6-filtered-tss")
s4.1.7.count.per.tss.dir <- file.path(s4.1.tss.def.dir, "s4-1-7-count-per-tss") 
s4.3.tx.info.dir <- file.path(s4.tss.dir, "s4-3-transcript-info")
s4.3.1.tx.info.rcc4.dir <- file.path(s4.3.tx.info.dir, "s4-3-1-transcript-info-for-RCC4")

s8.dir <- file.path(results.dir, "s8-analysis-of-translation")
s8.1.dir <- file.path(s8.dir, "s8-1-differentially-translated-mRNAs")
s8.1.1.dir <- file.path(s8.1.dir, "gene-level-dte")
s8.1.2.dir <- file.path(s8.1.dir, "tx-level-dte")
s8.3.dir <- file.path(s8.dir, "s8-3-validation-of-method")

sample.dt <- fread(sample.file)
sample.names <- sample.dt[, sample_name]

```

# Evaluation of translation changes by mTOR inhibition (gene level)

## VHL status for the analysis


```{r vhl_status}

vhl.status <- "VHL"

```


## Data import

```{r evaluation_of_translation_changes_by_mTOR_inhibition}

all.filtered.gene.dt <- file.path(
    s8.3.dir,
    "filtered_gene_for_polysome_analysis.csv"
) %>% fread


all.filtered.gene.dt[RCC4_VHL_NA == TRUE & RCC4_VHL_Torin1 == TRUE][
    gene_id %in%
    primary.tx.dt[!duplicated(gene_id)][biotype == "protein_coding", gene_id]    
] %>%
    nrow

```



```{r read_translation_analysis_data_by_gene_level}

torin.gene.trsl.dt <- file.path(
    s8.1.1.dir,
    paste0(
        "RCC4_",
        vhl.status,
        "_EIF4E2_yy_xx__Torin1_vs_NA.csv"
    )
) %>% fread

torin.gene.trsl.dt[
  , trsl_reg_class := case_when(
        translational_regulation %in% "Up" ~ "Preserved",
        translational_regulation %in% "Down" ~ "Repressed",
        TRUE ~ "Not significant"
    ) %>%
        factor(levels = c("Preserved", "Not significant", "Repressed"))
]

```

## Analysis of the absolute changes in translation upon mTOR inhibition


```{r plot_differential_translation_data}

torin.gene.trsl.dt <- torin.gene.trsl.dt[
    gene_id %in% all.filtered.gene.dt[RCC4_VHL_NA == TRUE & RCC4_VHL_Torin1 == TRUE, gene_id] &
    gene_id %in% primary.tx.dt[!duplicated(gene_id)][biotype == "protein_coding", gene_id]
]

ggplot(
    data = torin.gene.trsl.dt,
    aes(
        x = MRL_base,
        y = MRL_treated
    )
) +
    geom_abline(slope = 1, intercept = 0, color = "gray60") +
    geom_point() +
    xlab("MRL without Torin 1") +
    ylab("MRL with Torin 1") +
    xlim(c(1, 8)) + ylim(c(1, 8)) +
    theme(
        aspect.ratio = 1,
        legend.position = "bottom"
    )


```


# Analysis of changes in translational efficiency as a function of the functional classes of mRNAs


```{r torin_by_class, fig.height = 6, fig.width = 7}

## Plot translation changes by mRNA classes
plotTrslChangeByClass <- function(trsl.dt, s8.3.dir, plot.ylab, sig.th = 0.05){
    ## trsl.dt must have 2 columns: gene_id and MRL_log2fc

    trsl.dt <- trsl.dt[!is.na(MRL_log2fc)]

    kegg.all.dt <- file.path(s8.3.dir, "key_KEGG_genes.csv") %>%
        fread

    all.sl.terms <- kegg.all.dt[, unique(term_name)]
    all.sl.term.groups <- kegg.all.dt[, unique(term_group)]

    mrna.class.trsl.dt <- merge(
        kegg.all.dt,
        trsl.dt,
        by = "gene_id"
    )

    all.trsl.dt <- copy(trsl.dt) %>%
        {.[, `:=`(
             term_name = "All",
             term_group = "All"
         )]}

    mrna.class.trsl.dt <- rbind(
        mrna.class.trsl.dt,
        all.trsl.dt,
        use.names = TRUE
    )

    calcWilP <- function(sl.term, mrna.class.trsl.dt){
        dt <- data.table(
            term_name = sl.term,
            p_value = wilcox.test(
                mrna.class.trsl.dt[term_name == sl.term, MRL_log2fc],
                mrna.class.trsl.dt[
                    term_name == "All" &
                    !(gene_id %in% mrna.class.trsl.dt[term_name == sl.term, gene_id]),
                    MRL_log2fc
                ],
                alternative = "two.sided"
            )$p.value,
            rg =
                mrna.class.trsl.dt[order(
                    term_name == "All"
                )][term_name %in% c(sl.term, "All")][
                    !duplicated(gene_id)
                ] %$%
                rcompanion::wilcoxonRG(
                                x = MRL_log2fc,
                                g = term_name == "All"
                            ),
            N = nrow(mrna.class.trsl.dt[term_name == sl.term]),
            N_all = nrow(mrna.class.trsl.dt[term_name == "All"])
        )
        
        return(dt)
    }

    all.term.sig.dt <- lapply(
        all.sl.terms,
        calcWilP,
        mrna.class.trsl.dt = mrna.class.trsl.dt
    ) %>%
        rbindlist

    all.term.sig.dt[, `:=`(
        padj = p.adjust(p_value, method = "holm")
    )]

    all.term.sig.dt[, `:=`(
        direction = case_when(
            padj < sig.th & rg > 0 ~ "Up",
            padj < sig.th & rg < 0 ~ "Down",
            TRUE ~ "N.S."
        ),
        sig_mark = case_when(
            padj < sig.th * 0.1 ~ "**",
            padj < sig.th ~ "*",
            TRUE ~ NA_character_
        )
    )]
    
    print(all.term.sig.dt)

    all.term.sig.dt <- rbind(
        all.term.sig.dt,
        data.table(
            term_name = "All", direction = "All",
            N = nrow(mrna.class.trsl.dt[term_name == "All"])
        ),
        use.names = TRUE, fill = TRUE
    )

    all.term.sig.dt[, term_name_n := paste0(
                          term_name, " (", N, ")"
                      )]
    
    mrna.class.trsl.dt <- merge(
        mrna.class.trsl.dt,
        all.term.sig.dt,
        by = "term_name"
    )

    mrna.class.trsl.dt[, `:=`(
        term_name = factor(term_name, levels = c(all.sl.terms, "All")),
        term_group = factor(term_group, levels = c(all.sl.term.groups, "All")),
        term_name_n = factor(term_name_n, levels = all.term.sig.dt[, term_name_n])
    )]
    
    g1 <- ggplot(
        data = mrna.class.trsl.dt,
        aes(
            x = term_name_n,
            y = MRL_log2fc,
            color = direction,
            fill = direction
        )
    ) +
        geom_hline(
            yintercept = median(trsl.dt[, MRL_log2fc], na.rm = TRUE)
        ) +
        geom_boxplot(outlier.shape = NA) +
        stat_summary(
            geom = 'text', aes(label = sig_mark),
            fun = function(x){boxplot.stats(x)$stats[5]}, 
            vjust = -0.8, color = "black", size = 5
        ) +
        facet_grid(
            ~ term_group,
            scales = "free",
            space = "free"
        ) +
        scale_color_manual(
            values = c(
                "Up" = "#4477AA", "Down" = "#EE6677",
                "N.S." = "gray40", "All" = "black"
            )
        ) +
        scale_fill_manual(
            values = c(
                "Up" = "lightsteelblue2", "Down" = "mistyrose",
                "N.S." = "white", "All" = "white"
            )
        ) +
        xlab("Functional class") +
        ylab(plot.ylab) +
        theme(
            legend.position = "bottom"
        ) +
        scale_x_discrete(guide = guide_axis(angle = 90))

    print(g1)

    return(all.term.sig.dt)
}


all.term.sig.dt <- plotTrslChangeByClass(
    trsl.dt = torin.gene.trsl.dt,
    s8.3.dir = s8.3.dir,
    plot.ylab = "MRL log2 fold change with Torin 1",
    sig.th = 0.05
)

```

# Translational targets of mTOR defined by previous studies


Known mTOR hypersensitive classes were defined based on previous genome-wide studies.


## Map mouse genes in the data by Thoreen et al. to human genes


```{r map_mouse_genes_to_human_genes}

library("readxl")
library("gprofiler2")
set_base_url("https://biit.cs.ut.ee/gprofiler_archive3/e102_eg49_p15") # For reproducibility, use archive version

Thoreen.all.dt <- read_excel("../../data/others/PMID22552098_ST1.xls", skip = 2) %>%
    data.table

Thoreen.key.cols <- c("refseq_id", "trsl_eff_log2fc")

setnames(
    Thoreen.all.dt,
    old = c("Refseq Accession", "WT Efficiency: log2(Torin1/Vehicle)"),
    new = Thoreen.key.cols
)

Thoreen.all.dt <- Thoreen.all.dt[, Thoreen.key.cols, with = FALSE]

topN.down.refseq_ids <- Thoreen.all.dt[
    order(trsl_eff_log2fc)
][, head(.SD, n = 253)][, refseq_id]

topN.up.refseq_ids <- Thoreen.all.dt[
    order(trsl_eff_log2fc, decreasing = TRUE)
][, head(.SD, n = 198)][, refseq_id]

Thoreen.all.dt[
  , translational_regulation := case_when(
        refseq_id %in% topN.down.refseq_ids ~ "Down",
        refseq_id %in% topN.up.refseq_ids ~ "Up",
        TRUE ~ "Unclassified"
    )
]

ms2hs.dt <- gorth(
    query = Thoreen.all.dt[, refseq_id],
    source_organism = "mmusculus", 
    target_organism = "hsapiens",
    mthreshold = Inf, filter_na = TRUE,
    numeric_ns = "ENTREZGENE_ACC"    
) %>%
    data.table

setnames(
    ms2hs.dt,
    old = c("input", "input_ensg", "ortholog_ensg"),
    new = c("refseq_id", "ms_gene_id", "gene_id")
)

ms2hs.dt <- ms2hs.dt %>%
    group_by(refseq_id) %>%
    filter(n() == 1) %>%
    ungroup %>%
    group_by(ms_gene_id) %>%
    filter(n() == 1) %>%
    ungroup %>%
    group_by(gene_id) %>%
    filter(n() == 1) %>%
    data.table

Thoreen.all.dt <- merge(
    ms2hs.dt[, c("refseq_id", "ms_gene_id", "gene_id"), with = FALSE],
    Thoreen.all.dt,
    by = "refseq_id"
)

fwrite(Thoreen.all.dt, file.path(s8.3.dir, "Thoreen_et_al_with_human_gene_id.csv"))

```


## Other data


```{r class_of_mTOR_targets}

known.mtor.target.dt <- file.path(
    "../../data/others/20201127_previously_reported_mTOR_target_genes.csv"
) %>%
    fread

known.mtor.target.dt <- known.mtor.target.dt[
    (method == "Ribosome profiling"  & reported_gene_name != "C3ORF38") |
    (method == "Polysome profiling" & PP242_log2fc < -log2(1.5) & PP242_FDR < 0.15)
][, .(gene_id, method)]

Thoreen.all.dt[, method := paste0("RP_Thoreen_", translational_regulation)]

known.mtor.target.dt <- rbind(
    known.mtor.target.dt,
    copy(Thoreen.all.dt)[
        , method := paste0("RP_Thoreen_", translational_regulation)
    ][translational_regulation != "Unclassified", .(gene_id, method)],
    use.names = TRUE, fill = TRUE
)

kegg.all.dt <- file.path(s8.3.dir, "key_KEGG_genes.csv") %>%
    fread

kegg.all.dt[, term_count := .N, by = term_name]

known.mtor.target.kegg.dt <- merge(
    kegg.all.dt,
    known.mtor.target.dt[, .(gene_id, method)],
    by = "gene_id"
)

known.mtor.target.kegg.dt[, target_count := .N, by = list(term_name, method)]

target.count.per.term.dt <- known.mtor.target.kegg.dt[
    !duplicated(paste(term_name, method)),
    .(term_group, term_name, method, target_count, term_count)
]

target.count.per.term.dt <- merge(
    target.count.per.term.dt,
    all.term.sig.dt[, .(term_name, N)],
    by = "term_name"
)

target.count.per.term.dt[, `:=`(
    target_ratio = target_count / term_count,
    target_ratio_to_N_RCC4VHL = target_count / N
)]

target.count.per.term.dt[, `:=`(
    systematic_flag = target_ratio >= 0.1
)]

target.count.per.term.dt[
    order(term_name, method, -target_ratio_to_N_RCC4VHL)
][systematic_flag == TRUE]


```


# Session information

```{r sessionInfo}

sessionInfo()

```
